buildSrc/src/main/groovy/org/gradle/build/docs/ExtractSnippetsTask.groovy

```diff
@@ -15,38 +15,67 @@
  */
 package org.gradle.build.docs
 
+import org.gradle.api.Action
+import org.gradle.api.DefaultTask
 import org.gradle.api.file.FileTree
 import org.gradle.api.file.FileVisitDetails
 import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputFiles
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.OutputDirectory
 import org.gradle.api.tasks.PathSensitive
 import org.gradle.api.tasks.PathSensitivity
-import org.gradle.api.tasks.OutputDirectory
-import org.gradle.api.tasks.SourceTask
 import org.gradle.api.tasks.TaskAction
+import org.gradle.api.tasks.util.PatternFilterable
 
 import java.util.regex.Matcher
 import java.util.regex.Pattern
+
 /**
  * Produces the snippets files for a set of sample source files.
  */
 @CacheableTask
-class ExtractSnippetsTask extends SourceTask {
+class ExtractSnippetsTask extends DefaultTask {
 
     @OutputDirectory
     File destDir
 
     @OutputDirectory
     File snippetsDir
 
-    @Override
+    @Input
+    List<String> excludes
+
+    // Resources that should not be filtered
+    @Input
+    List<String> nonFiltered
+
+    @Internal
+    File samples
+
+    @InputFiles
     @PathSensitive(PathSensitivity.RELATIVE)
     FileTree getSource() {
-        return super.getSource()
+        def source = project.fileTree(samples)
+        source.excludes = excludes
+        return source
     }
 
     @TaskAction
     def extract() {
-        source.visit { FileVisitDetails details ->
+        project.copy { copySpec ->
+            copySpec.from samples
+            copySpec.into destDir
+            copySpec.setIncludes(nonFiltered)
+        }
+
+        getSource().matching(new Action<PatternFilterable>() {
```

Comment:
You could use a `Closure` instead of an `Action<PatternFilterable>`. What are your reasons to go with the `Action`?

Reply:
Auto-complete guided me here.  I've take a different stab at it now